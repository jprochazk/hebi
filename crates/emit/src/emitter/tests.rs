use super::*;

macro_rules! check {
  ($input:literal) => {{
    let input = indoc::indoc!($input);
    let module = match syntax::parse(input) {
      Ok(module) => module,
      Err(e) => {
        for err in e {
          eprintln!("{}", err.report(input));
        }
        panic!("Failed to parse source, see errors above.")
      }
    };
    let ctx = Context::new();
    let chunk = match emit(ctx, "[[main]]", &module) {
      Ok(chunk) => chunk,
      Err(e) => {
        panic!("failed to emit chunk:\n{}", e.report(input));
      }
    };
    let snapshot = format!(
      "# Input:\n{input}\n\n# Chunk:\n{}",
      chunk.disassemble(op::instruction::disassemble, false)
    );
    insta::assert_snapshot!(snapshot);
  }};
}

#[test]
fn print_literals() {
  check!(r#"print 0"#);
  check!(r#"print 2.5"#);
  check!(r#"print "test""#);
  check!(r#"print [0, 1, 2]"#);
  check!(r#"print { a: 0, b: 1, c: 2 }"#);
}

#[test]
fn print_variable() {
  check! {
    r#"
      v := 0
      print v # load_reg
    "#
  }
  // TODO: capture + nested capture
  check! {
    r#"
      print v # load_global
    "#
  }
}

#[test]
fn print_field() {
  check! {
    r#"
      v := { a: 0 }
      print v.a
      v.a = 1
    "#
  }
  check! {
    r#"
      v := { a: 0 }
      print v["a"]
      v["a"] = 1
    "#
  }
  check! {
    r#"
      v := {}
      print ?v.a
      print ?v.a.b.c
    "#
  }
  check! {
    r#"
      v := {}
      print ?v["a"]
      print ?v["a"]["b"]["c"]
    "#
  }
}

#[test]
fn call() {
  check!(r#"f()"#);
  check!(r#"f(0)"#);
  check!(r#"f(0, 1, 2)"#);
  check!(r#"f(a=0)"#);
  check!(r#"f(a=0, b=1, c=2)"#);
  check!(r#"f(a, b, c=2)"#);
  check!(r#"a(b(c()))"#);
}

#[test]
fn func() {
  check! {
    r#"
      fn test():
        pass

      test()
    "#
  }

  check! {
    r#"
      fn test(a):
        print a
      
      test(0)
    "#
  }

  check! {
    r#"
      fn test(a, b=10):
        print a, b

      test(1)
      test(1, 2)
    "#
  }

  check! {
    r#"
      fn test(a, *, b):
        print a, b

      test(1, b=2)
    "#
  }

  check! {
    r#"
      fn test(a, *, b=10):
        print a, b

      test(1)
      test(1, b=2)
    "#
  }

  check! {
    r#"
      fn test(a, *v, b=10, **kw):
        print a, v, b, kw

      test(1, 2)
      test(1, 2, b=3, c=4)
    "#
  }
}

#[test]
fn closure() {
  check! {
    r#"
      fn a():
        v := 0
        fn b():
          print v
        return b
      
      a()()
    "#
  }

  check! {
    r#"
      fn a():
        v := 0
        fn b():
          fn c():
            fn d():
              print v
    "#
  }
}

#[test]
fn logical_expr() {
  check! {
    r#"
      fn f0(a, b):
        a && b
        a || b
        a ?? b


      fn f1_a(a, b, c, d):
        (a && b) || (c && d)
      fn f1_b(a, b, c, d):
         a && b  ||  c && d
      
      fn f2_a(a, b, c, d):
        (a || (b && c)) || d
      fn f2_b(a, b, c, d):
        a ||  b && c  || d

      fn f3(a, b, c, d):
        a ?? b ?? c
    "#
  }
}

#[test]
fn if_stmt() {
  check! {
    r#"
      if true:
        print a
      elif true:
        print b
      else:
        print c
    "#
  }

  check! {
    r#"
      if a:
        b := a
        print b
      else:
        print b
    "#
  }
}
