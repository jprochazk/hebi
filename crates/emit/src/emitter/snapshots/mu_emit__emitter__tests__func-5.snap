---
source: crates/emit/src/emitter/tests.rs
expression: snapshot
---
# Input:
fn test(a, *, b=10):
  print a, b

test(1)
test(1, b=2)


# Chunk:
function <test>:
  frame_size: 6
  length: 33
  const (length=1):
    0: "b"
  code:
     0 | 2f 00             is_kw_param_not_set [0]
     2 | 19 0d             jump_if_false offset=13
     4 | 00                nop 
     5 | 00                nop 
     6 | 00                nop 
     7 | 00                nop 
     8 | 11 0a 00 00 00    push_small_int value=10
    13 | 05 05             store_reg r5
    15 | 30 00 05          load_kw_param [0], r5
    18 | 12                create_empty_list 
    19 | 05 00             store_reg r0
    21 | 04 04             load_reg r4
    23 | 13 00             push_to_list r0
    25 | 04 05             load_reg r5
    27 | 13 00             push_to_list r0
    29 | 2b 00             print_list r0
    31 | 31                ret 
    32 | ff                suspend 

function <test>:
  frame_size: 3
  length: 45
  const (length=5):
    0: <func test>
    1: "test"
    2: "test"
    3: "test"
    4: "b"
  code:
     0 | 03 00             load_const [0]
     2 | 09 01             store_global [1]
     4 | 08 02             load_global [2]
     6 | 05 00             store_reg r0
     8 | 11 01 00 00 00    push_small_int value=1
    13 | 05 01             store_reg r1
    15 | 2c 00 01          call r0, args=1
    18 | 08 03             load_global [3]
    20 | 05 00             store_reg r0
    22 | 14                create_empty_dict 
    23 | 05 01             store_reg r1
    25 | 11 01 00 00 00    push_small_int value=1
    30 | 05 02             store_reg r2
    32 | 11 02 00 00 00    push_small_int value=2
    37 | 16 04 01          insert_to_dict_keyed [4], r1
    40 | 2d 00 01          call_kw r0, args=1
    43 | 31                ret 
    44 | ff                suspend 

